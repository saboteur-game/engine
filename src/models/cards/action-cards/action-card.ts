import Card from "../card";
import CardParameters from "../card-parameters";
import { Pojo } from "../../../utils";

class ActionCard extends Card {
  parameters: CardParameters | undefined;

  play(parameters: CardParameters): ActionCard {
    super.play(parameters);
    this.parameters = parameters;
    return this;
  }

  toJSON(): Pojo {
    return {
      ...super.toJSON(),
      parameters: this.parameters,
    };
  }
}

export default ActionCard;
