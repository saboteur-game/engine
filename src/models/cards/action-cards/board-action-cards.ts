import { ActionCard } from ".";
import CardParameters from "../card-parameters";

export class BoardActionCard extends ActionCard {
  play(parameters: CardParameters): BoardActionCard {
    return super.play(parameters);
  }
}

export class MapActionCard extends BoardActionCard {}

export class RockfallActionCard extends BoardActionCard {}
