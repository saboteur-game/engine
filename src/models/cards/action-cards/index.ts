import ActionCard from "./action-card";
import {
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
} from "./board-action-cards";
import {
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
} from "./tool-action-cards";

export {
  ActionCard,
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
};
