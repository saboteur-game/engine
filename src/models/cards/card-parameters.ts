import { Tools } from "../../constants";
import Position from "../position";
import Player from "../player";

export interface IBoardCardParameters {
  position: Position;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
