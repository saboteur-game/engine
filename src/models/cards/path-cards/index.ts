import PathCard from "./path-card";
import { TunnelCard, PassageCard, DeadendCard } from "./tunnel-cards";
import {
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
} from "./boundary-cards";

export {
  PathCard,
  TunnelCard,
  PassageCard,
  DeadendCard,
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
};
