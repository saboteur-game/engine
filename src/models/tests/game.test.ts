import { mocked } from "ts-jest/utils";
import eventEmitter from "../../event-emitter";
import Player from "../player";
import { PathCard } from "../cards/path-cards";
import { MapActionCard, ActionCard } from "../cards/action-cards";
import CardParameters from "../cards/card-parameters";
import { getShuffledDeck } from "../cards";
import { middleFinishPosition } from "../board";
import Game from "../game";

jest.mock("../../event-emitter");
jest.mock("../cards", () => {
  jest.spyOn(global.Math, "random").mockReturnValue(0.5);
  const cards = jest.requireActual("../cards");
  return {
    ...cards,
    getShuffledDeck: jest.fn().mockReturnValue(cards.getShuffledDeck()),
  };
});

const mockedGetSuffledDeck = mocked(getShuffledDeck);

describe("Game", () => {
  let game: Game;

  const addPlayersToGame = (count: number): Player[] => {
    const players = Array(count)
      .fill(undefined)
      .map((value, index) => new Player(`Player ${index + 1}`, 20 + index));
    players.forEach((player) => game.addPlayer(player));
    return players;
  };

  const runForEmitterTest = (wrappedAction: () => void) => {
    jest.clearAllMocks();
    try {
      wrappedAction();
    } catch (e) {
      // noop
    }
  };

  beforeEach(() => {
    jest.clearAllMocks();
    game = new Game();
  });

  it("can be serialized", () => {
    expect(game.toJS()).toMatchSnapshot();
  });

  it("can visualize the board", () => {
    expect(game.visualizeBoard()).toMatchSnapshot();
  });

  describe("on event", () => {
    const mockEventHandler = jest.fn();

    beforeEach(() => {
      game.on("*", mockEventHandler);
    });

    it("sets up event listener", () => {
      expect(eventEmitter.on).toHaveBeenCalledTimes(1);
      expect(eventEmitter.on).toHaveBeenCalledWith("*", mockEventHandler);
    });
  });

  describe("addPlayer", () => {
    let player: Player;

    beforeEach(() => {
      player = new Player("Alice", 35);
    });

    describe("when maximum number of players have been added", () => {
      beforeEach(() => {
        addPlayersToGame(10);
        jest.clearAllMocks();
      });

      it("throws exception", () => {
        expect(() => game.addPlayer(player)).toThrow(
          "Maximum number of players"
        );
        expect(game.getPlayers().length).toBe(10);
      });

      it("does not emit event", () => {
        runForEmitterTest(() => game.addPlayer(player));
        expect(eventEmitter.emit).not.toHaveBeenCalled();
      });
    });

    describe("when no players have been added", () => {
      beforeEach(() => {
        game.addPlayer(player);
      });

      it("adds the player", () => {
        expect(game.getPlayers()).toMatchSnapshot();
      });

      it("emits add-player event", () => {
        expect(eventEmitter.emit).toHaveBeenCalledTimes(1);
        expect(eventEmitter.emit).toHaveBeenCalledWith("add-player", player);
      });
    });
  });

  describe("removePlayer", () => {
    let player: Player;

    beforeEach(() => {
      player = new Player("Alice", 35);
    });

    describe("when there are no players added", () => {
      it("does nothing", () => {
        expect(game.getPlayers().length).toBe(0);
        game.removePlayer(player.id);
        expect(game.getPlayers().length).toBe(0);
      });

      it("does not emit event", () => {
        jest.clearAllMocks();
        game.removePlayer(player.id);
        expect(eventEmitter.emit).not.toHaveBeenCalled();
      });
    });

    describe("when players have been added", () => {
      beforeEach(() => {
        game.addPlayer(player);
        jest.clearAllMocks();
      });

      it("removes the player", () => {
        expect(game.getPlayers().length).toBe(1);
        game.removePlayer(player.id);
        expect(game.getPlayers().length).toBe(0);
      });

      it("emits remove-player event", () => {
        game.removePlayer(player.id);
        expect(eventEmitter.emit).toHaveBeenCalledTimes(1);
        expect(eventEmitter.emit).toHaveBeenCalledWith("remove-player", player);
      });
    });
  });

  describe("start", () => {
    describe("without enough players", () => {
      beforeEach(() => {
        addPlayersToGame(2);
      });

      it("throws exception", () => {
        expect(() => game.start()).toThrow("Not enough players");
      });

      it("does not emit event", () => {
        runForEmitterTest(() => game.start());
        expect(eventEmitter.emit).not.toHaveBeenCalled();
      });
    });

    describe("with enough players", () => {
      let players: Player[];

      beforeEach(() => {
        players = addPlayersToGame(3);
      });

      it("sets player allegiances and deals cards", () => {
        game.start();
        expect(game.getPlayers()).toMatchSnapshot();
      });

      it("emits start-game and start-turn events", () => {
        jest.clearAllMocks();
        game.start();
        expect(eventEmitter.emit).toHaveBeenCalledTimes(2);
        expect(eventEmitter.emit).toHaveBeenCalledWith("start-game");
        expect(eventEmitter.emit).toHaveBeenCalledWith(
          "start-turn",
          players[0]
        );
      });
    });
  });

  describe("getActivePlayer", () => {
    let players: Player[];

    beforeEach(() => {
      players = addPlayersToGame(3);
    });

    describe("when game is not started", () => {
      it("returns undefined", () => {
        expect(game.getActivePlayer()).toBe(undefined);
      });
    });

    describe("when game is started", () => {
      beforeEach(() => {
        game.start();
      });

      it("returns undefined", () => {
        expect(game.getActivePlayer()).toBe(players[0]);
      });
    });
  });

  describe("getPlayer", () => {
    let player: Player;

    beforeEach(() => {
      player = new Player("Alice", 35);
    });

    describe("when there are no players added", () => {
      it("returns undefined", () => {
        expect(game.getPlayer(player.id)).toBe(undefined);
      });
    });

    describe("when players have been added", () => {
      beforeEach(() => {
        game.addPlayer(player);
      });

      it("get the player", () => {
        expect(game.getPlayer(player.id)).toBe(player);
      });
    });
  });

  describe("playCard", () => {
    let players: Player[];
    let card: MapActionCard;
    let parameters: CardParameters;

    beforeEach(() => {
      players = addPlayersToGame(3);
      card = new MapActionCard();
      parameters = { position: middleFinishPosition };
    });

    describe("when game is not started", () => {
      it("throws exception", () => {
        expect(() => game.playCard(players[0], card.id, parameters)).toThrow(
          "Game has not started"
        );
      });

      it("does not emit event", () => {
        runForEmitterTest(() => game.playCard(players[0], card.id, parameters));
        expect(eventEmitter.emit).not.toHaveBeenCalled();
      });
    });

    describe("when game is started", () => {
      beforeEach(() => {
        game.start();
      });

      describe("card is not in the players hand", () => {
        it("throws exception", () => {
          expect(() => game.playCard(players[0], card.id, parameters)).toThrow(
            "Cannot play card which isn't in players hand"
          );
        });

        it("does not emit event", () => {
          runForEmitterTest(() =>
            game.playCard(players[0], card.id, parameters)
          );
          expect(eventEmitter.emit).not.toHaveBeenCalled();
        });
      });

      describe("card is in the players hand", () => {
        beforeEach(() => {
          players[0].addToHand(card);
          jest.clearAllMocks();
          game.playCard(players[0], card.id, parameters);
        });

        it("adds a new card to the players hand", () => {
          expect(players[0].getHand().includes(card)).toBe(false);
          expect(players[0].getHand()).toMatchSnapshot();
        });

        it("adds the card to the discard pile", () => {
          expect(game.getTopOfDiscardPile()).toMatchSnapshot();
        });

        it("ends the players turn and starts the next players turn", () => {
          expect(game.getActivePlayer()).toBe(players[1]);
        });

        it("emits play-card, end-turn and start-turn events", () => {
          expect(eventEmitter.emit).toHaveBeenCalledTimes(3);
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "play-card",
            players[0],
            card
          );
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "end-turn",
            players[0]
          );
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "start-turn",
            players[1]
          );
        });
      });
    });
  });

  describe("discardCard", () => {
    let players: Player[];
    let card: MapActionCard;

    beforeEach(() => {
      players = addPlayersToGame(3);
      card = new MapActionCard();
    });

    describe("when game is not started", () => {
      it("throws exception", () => {
        expect(() => game.discardCard(players[0], card.id)).toThrow(
          "Game has not started"
        );
      });

      it("does not emit event", () => {
        runForEmitterTest(() => game.discardCard(players[0], card.id));
        expect(eventEmitter.emit).not.toHaveBeenCalled();
      });
    });

    describe("when game is started", () => {
      beforeEach(() => {
        game.start();
      });

      describe("card is not in the players hand", () => {
        it("throws exception", () => {
          expect(() => game.discardCard(players[0], card.id)).toThrow(
            "Cannot discard card which isn't in players hand"
          );
        });

        it("does not emit event", () => {
          runForEmitterTest(() => game.discardCard(players[0], card.id));
          expect(eventEmitter.emit).not.toHaveBeenCalled();
        });
      });

      describe("no card is supplied", () => {
        describe("when the player has cards in their hand", () => {
          beforeEach(() => {
            players[0].addToHand(card);
            jest.clearAllMocks();
          });

          it("throws exception", () => {
            expect(() => game.discardCard(players[0])).toThrow(
              "Player must discard"
            );
          });

          it("does not emit event", () => {
            runForEmitterTest(() => game.discardCard(players[0]));
            expect(eventEmitter.emit).not.toHaveBeenCalled();
          });
        });

        describe("when the player has no cards in their hand", () => {
          beforeEach(() => {
            players[0]
              .getHand()
              .forEach((card: ActionCard | PathCard) =>
                players[0].removeFromHand(card.id)
              );
            jest.clearAllMocks();
            game.discardCard(players[0]);
          });

          it("adds the card to the discard pile", () => {
            expect(game.getTopOfDiscardPile()).toBe(undefined);
          });

          it("does not add cards to the players hand", () => {
            expect(players[0].getHandCardCount()).toBe(0);
          });

          it("ends the players turn and starts the next players turn", () => {
            expect(game.getActivePlayer()).toBe(players[1]);
          });

          it("emits discard-card, end-turn and start-turn events", () => {
            expect(eventEmitter.emit).toHaveBeenCalledTimes(3);
            expect(eventEmitter.emit).toHaveBeenCalledWith(
              "discard-card",
              players[0],
              null
            );
            expect(eventEmitter.emit).toHaveBeenCalledWith(
              "end-turn",
              players[0]
            );
            expect(eventEmitter.emit).toHaveBeenCalledWith(
              "start-turn",
              players[1]
            );
          });
        });
      });

      describe("card is in the players hand", () => {
        beforeEach(() => {
          players[0].addToHand(card);
          jest.clearAllMocks();
          game.discardCard(players[0], card.id);
        });

        it("adds a new card to the players hand", () => {
          expect(players[0].getHand().includes(card)).toBe(false);
          expect(players[0].getHand()).toMatchSnapshot();
        });

        it("adds the card to the discard pile", () => {
          expect(game.getTopOfDiscardPile()).toBe(null);
        });

        it("ends the players turn and starts the next players turn", () => {
          expect(game.getActivePlayer()).toBe(players[1]);
        });

        it("emits discard-card, end-turn and start-turn events", () => {
          expect(eventEmitter.emit).toHaveBeenCalledTimes(3);
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "discard-card",
            players[0],
            card
          );
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "end-turn",
            players[0]
          );
          expect(eventEmitter.emit).toHaveBeenCalledWith(
            "start-turn",
            players[1]
          );
        });
      });
    });
  });

  describe("when the deck is reduced", () => {
    let players: Player[];

    beforeEach(() => {
      jest.clearAllMocks();
      mockedGetSuffledDeck.mockReturnValue([
        new MapActionCard(),
        new MapActionCard(),
        new MapActionCard(),
      ]);
      game = new Game();
      players = addPlayersToGame(3);
      game.start();
    });

    it("deals available cards to players", () => {
      expect(players[0].getHand().length).toBe(1);
      expect(players[1].getHand().length).toBe(1);
      expect(players[2].getHand().length).toBe(1);
    });

    describe("playCard", () => {
      beforeEach(() => {
        const card = players[0].getHand()[0];
        const parameters = { position: middleFinishPosition };
        game.playCard(players[0], card.id, parameters);
      });

      it("does not add a new card to the players hand", () => {
        expect(players[0].getHand().length).toBe(0);
      });
    });

    describe("discardCard", () => {
      beforeEach(() => {
        const card = players[0].getHand()[0];
        game.discardCard(players[0], card.id);
      });

      it("does not add a new card to the players hand", () => {
        expect(players[0].getHand().length).toBe(0);
      });
    });
  });
});
