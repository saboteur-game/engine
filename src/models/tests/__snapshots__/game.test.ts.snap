// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game addPlayer when no players have been added adds the player 1`] = `
Array [
  Player {
    "age": 35,
    "hand": Object {},
    "id": "test-id-8",
    "isSaboteur": false,
    "name": "Alice",
    "tools": Object {
      "lamp": true,
      "pick": true,
      "wagon": true,
    },
    "viewedFinishCards": Object {},
  },
]
`;

exports[`Game can be serialized 1`] = `
Object {
  "board": Board {
    "grid": Object {
      "0,0": StartPathCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-4",
        "isUpsideDown": false,
        "status": "played",
      },
      "8,-2": RockFinishPathCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-7",
        "isFaceDown": true,
        "isUpsideDown": false,
        "status": "played",
      },
      "8,0": RockFinishPathCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-6",
        "isFaceDown": true,
        "isUpsideDown": false,
        "status": "played",
      },
      "8,2": GoldFinishPathCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-5",
        "isFaceDown": true,
        "isUpsideDown": false,
        "status": "played",
      },
    },
    "id": "test-id-3",
    "isComplete": false,
  },
  "deck": Deck {
    "drawPile": Array [
      PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-0",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-1",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-2",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-3",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-4",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-5",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-6",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-7",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-8",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-9",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-10",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-11",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-12",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-13",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-14",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-15",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-16",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-17",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-18",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-19",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-20",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-21",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-22",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-23",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-24",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-25",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-26",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-27",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-28",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-29",
        "isUpsideDown": false,
        "status": "unused",
      },
      PassageCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-30",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
        ],
        "id": "test-id-31",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          2,
        ],
        "id": "test-id-32",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-33",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-34",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-35",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-36",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          2,
          3,
        ],
        "id": "test-id-37",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-38",
        "isUpsideDown": false,
        "status": "unused",
      },
      DeadendCard {
        "connectors": Array [
          1,
          2,
          3,
          4,
        ],
        "id": "test-id-39",
        "isUpsideDown": false,
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-40",
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-41",
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-42",
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-43",
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-44",
        "status": "unused",
      },
      MapActionCard {
        "id": "test-id-45",
        "status": "unused",
      },
      RockfallActionCard {
        "id": "test-id-46",
        "status": "unused",
      },
      RockfallActionCard {
        "id": "test-id-47",
        "status": "unused",
      },
      RockfallActionCard {
        "id": "test-id-48",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "pick",
        ],
        "id": "test-id-49",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "pick",
        ],
        "id": "test-id-50",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "pick",
        ],
        "id": "test-id-51",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "lamp",
        ],
        "id": "test-id-52",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "lamp",
        ],
        "id": "test-id-53",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "lamp",
        ],
        "id": "test-id-54",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "wagon",
        ],
        "id": "test-id-55",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "wagon",
        ],
        "id": "test-id-56",
        "status": "unused",
      },
      BreakActionCard {
        "appliesTo": Array [
          "wagon",
        ],
        "id": "test-id-57",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "pick",
        ],
        "id": "test-id-58",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "pick",
        ],
        "id": "test-id-59",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "lamp",
        ],
        "id": "test-id-60",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "lamp",
        ],
        "id": "test-id-61",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "wagon",
        ],
        "id": "test-id-62",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "wagon",
        ],
        "id": "test-id-63",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "pick",
          "lamp",
        ],
        "id": "test-id-64",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "pick",
          "wagon",
        ],
        "id": "test-id-65",
        "status": "unused",
      },
      RepairActionCard {
        "appliesTo": Array [
          "lamp",
          "wagon",
        ],
        "id": "test-id-66",
        "status": "unused",
      },
    ],
    "id": "test-id-1",
  },
  "discard": Discard {
    "discardPile": Array [],
    "id": "test-id-2",
  },
  "id": "test-id-0",
  "isFinished": false,
  "isStarted": false,
  "playOrder": Array [],
  "players": Object {},
  "roundResults": Array [],
  "turn": 0,
}
`;

exports[`Game can visualize the board 1`] = `
"                                                      ▕▔▔▔▔▏
                                                      ▕ ?? ▏
                                                      ▕▁▁▁▁▏
      ╭┄┄┄┄╮                                                
      │    │                                                
      ╰┄┄┄┄╯                                                
╭┄┄┄┄╮██  ██╭┄┄┄┄╮                                    ▕▔▔▔▔▏
│    │  []  │    │                                    ▕ ?? ▏
╰┄┄┄┄╯██  ██╰┄┄┄┄╯                                    ▕▁▁▁▁▏
      ╭┄┄┄┄╮                                                
      │    │                                                
      ╰┄┄┄┄╯                                                
                                                      ▕▔▔▔▔▏
                                                      ▕ ?? ▏
                                                      ▕▁▁▁▁▏"
`;

exports[`Game discardCard when game is started card is in the players hand adds a new card to the players hand 1`] = `
Array [
  PassageCard {
    "connectors": Array [
      2,
      4,
    ],
    "id": "test-id-0",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      3,
    ],
    "id": "test-id-3",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      3,
    ],
    "id": "test-id-6",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      4,
    ],
    "id": "test-id-9",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
    ],
    "id": "test-id-12",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
    ],
    "id": "test-id-15",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
      4,
    ],
    "id": "test-id-18",
    "isUpsideDown": false,
    "status": "unused",
  },
]
`;

exports[`Game playCard when game is started card is in the players hand adds a new card to the players hand 1`] = `
Array [
  PassageCard {
    "connectors": Array [
      2,
      4,
    ],
    "id": "test-id-0",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      3,
    ],
    "id": "test-id-3",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      3,
    ],
    "id": "test-id-6",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      4,
    ],
    "id": "test-id-9",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
    ],
    "id": "test-id-12",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
    ],
    "id": "test-id-15",
    "isUpsideDown": false,
    "status": "unused",
  },
  PassageCard {
    "connectors": Array [
      1,
      2,
      4,
    ],
    "id": "test-id-18",
    "isUpsideDown": false,
    "status": "unused",
  },
]
`;

exports[`Game playCard when game is started card is in the players hand adds the card to the discard pile 1`] = `
MapActionCard {
  "id": "test-id-11",
  "parameters": Object {
    "position": Position {
      "id": "test-id-69",
      "x": 8,
      "y": 0,
    },
  },
  "status": "played",
}
`;

exports[`Game start with enough players sets player allegiances and deals cards 1`] = `
Array [
  Player {
    "age": 20,
    "hand": Object {
      "test-id-0": PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-0",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-12": PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-12",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-15": PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-15",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-3": PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-3",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-6": PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-6",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-9": PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-9",
        "isUpsideDown": false,
        "status": "unused",
      },
    },
    "id": "test-id-8",
    "isSaboteur": true,
    "name": "Player 1",
    "tools": Object {
      "lamp": true,
      "pick": true,
      "wagon": true,
    },
    "viewedFinishCards": Object {},
  },
  Player {
    "age": 21,
    "hand": Object {
      "test-id-1": PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-1",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-10": PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-10",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-13": PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-13",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-16": PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-16",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-4": PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-4",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-7": PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-7",
        "isUpsideDown": false,
        "status": "unused",
      },
    },
    "id": "test-id-9",
    "isSaboteur": false,
    "name": "Player 2",
    "tools": Object {
      "lamp": true,
      "pick": true,
      "wagon": true,
    },
    "viewedFinishCards": Object {},
  },
  Player {
    "age": 22,
    "hand": Object {
      "test-id-11": PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-11",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-14": PassageCard {
        "connectors": Array [
          1,
          2,
        ],
        "id": "test-id-14",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-17": PassageCard {
        "connectors": Array [
          1,
          2,
          4,
        ],
        "id": "test-id-17",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-2": PassageCard {
        "connectors": Array [
          2,
          4,
        ],
        "id": "test-id-2",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-5": PassageCard {
        "connectors": Array [
          1,
          3,
        ],
        "id": "test-id-5",
        "isUpsideDown": false,
        "status": "unused",
      },
      "test-id-8": PassageCard {
        "connectors": Array [
          1,
          4,
        ],
        "id": "test-id-8",
        "isUpsideDown": false,
        "status": "unused",
      },
    },
    "id": "test-id-10",
    "isSaboteur": false,
    "name": "Player 3",
    "tools": Object {
      "lamp": true,
      "pick": true,
      "wagon": true,
    },
    "viewedFinishCards": Object {},
  },
]
`;
